packages/language-server/src/completion-items.ts

```diff
@@ -27,41 +27,75 @@ import {
 } from "@ui5-language-assistant/xml-views-completion";
 import { ui5NodeToFQN } from "@ui5-language-assistant/logic-utils";
 import { getNodeDocumentation, getNodeDetail } from "./documentation";
+import { getSettingsForDocument } from "@ui5-language-assistant/settings";
 import { assertNever } from "assert-never";
 
-export function getCompletionItems(
-  model: UI5SemanticModel,
-  textDocumentPosition: TextDocumentPositionParams,
-  document: TextDocument
-): CompletionItem[] {
-  const documentText = document.getText();
+export async function getCompletionItems(opts: {
+  model: UI5SemanticModel;
+  textDocumentPosition: TextDocumentPositionParams;
+  document: TextDocument;
+}): Promise<CompletionItem[]> {
+  const documentText = opts.document.getText();
   const { cst, tokenVector } = parse(documentText);
   const ast = buildAst(cst as DocumentCstNode, tokenVector);
   const suggestions = getXMLViewCompletions({
-    model: model,
-    offset: document.offsetAt(textDocumentPosition.position),
+    model: opts.model,
+    offset: opts.document.offsetAt(opts.textDocumentPosition.position),
     cst: cst as DocumentCstNode,
     ast: ast,
     tokenVector: tokenVector,
   });
 
+  const filteredSuggestions = await filterBySettings(
```

Comment:
naively I would prefer to pass the settings to this function as a param
instead of reading the settings everytime.

This looks like part of the state the language server needs to maintain such as UI5 Version / manifest.json property for stable ID validation ect.

I don't think the functional approach of no-state is good here when we are dealing with potentially slow File System operations.

Nor do I think its the responsibility of the completionProvider to read the settings

Reply:
The settings are read from a cache, not from the client. (It's in a `Thenable` to make sure we don't set the settings for the same document twice at the same time because the first time is still being resolved, not because it has to be re-resolved every time). So it's not a file system operation (although the first time might wait once until the request is completed).

Reply:
Changed